File: src/game/scoring.ts
export function lerp(a: number, b: number, t: number): number {
  return a + (b - a) * t;
}

/** 0 = obvious, 1 = tiny. First rounds stay readable; later ones tighten. */
export function difficultyT(roundIndex: number): number {
  if (roundIndex <= 3) return roundIndex * 0.07;
  return Math.min(1, 0.28 + (1 - Math.exp(-(roundIndex - 3) / 11)) * 0.72);
}

export function roundDurationMs(roundIndex: number): number {
  const t = Math.min(1, roundIndex / 16);
  return Math.round(lerp(7200, 2400, t));
}

/**
 * combo is the streak after this hit (1 on first hit).
 * Multiplier 1.0, 1.2, 1.4 … capped at 3.0.
 */
export function comboMultiplier(combo: number): number {
  const n = Math.max(1, combo);
  return 1 + Math.min(n - 1, 10) * 0.2;
}

export function scoreHit(opts: {
  roundIndex: number;
  remainingMs: number;
  durationMs: number;
  combo: number;
}): number {
  const base = 100 + opts.roundIndex * 40;
  const frac = opts.durationMs <= 0 ? 0 : Math.max(0, Math.min(1, opts.remainingMs / opts.durationMs));
  const timeBonus = Math.round(frac * (50 + opts.roundIndex * 8));
  return Math.round((base + timeBonus) * comboMultiplier(opts.combo));
}

export function scoreMiss(): number {
  return 0;
}

export function formatScore(n: number): string {
  return Math.max(0, Math.round(n)).toLocaleString("en-US");
}

export function formatCountdown(ms: number): string {
  const total = Math.max(0, Math.floor(ms / 1000));
  const h = Math.floor(total / 3600);
  const m = Math.floor((total % 3600) / 60);
  if (h > 0) return `${h}h ${m}m`;
  const s = total % 60;
  if (m > 0) return `${m}m ${s}s`;
  return `${s}s`;
}

File: android/app/src/main/java/app/skew/android/game/Scoring.kt
package app.skew.android.game

import java.util.Locale
import kotlin.math.exp
import kotlin.math.floor
import kotlin.math.roundToInt

fun lerp(a: Double, b: Double, t: Double): Double = a + (b - a) * t

fun difficultyT(roundIndex: Int): Double {
    if (roundIndex <= 3) return roundIndex * 0.07
    return minOf(1.0, 0.28 + (1 - exp(-(roundIndex - 3) / 11.0)) * 0.72)
}

fun roundDurationMs(roundIndex: Int): Int {
    val t = minOf(1.0, roundIndex / 16.0)
    return lerp(7200.0, 2400.0, t).roundToInt()
}

fun comboMultiplier(combo: Int): Double {
    val n = maxOf(1, combo)
    return 1.0 + minOf(n - 1, 10) * 0.2
}

fun scoreHit(roundIndex: Int, remainingMs: Double, durationMs: Double, combo: Int): Int {
    val base = 100.0 + roundIndex * 40
    val frac =
        if (durationMs <= 0.0) 0.0
        else remainingMs.coerceIn(0.0, durationMs) / durationMs
    val timeBonus = (frac * (50 + roundIndex * 8)).roundToInt()
    return ((base + timeBonus) * comboMultiplier(combo)).roundToInt()
}

fun scoreMiss(): Int = 0

fun formatScore(n: Int): String =
    String.format(Locale.US, "%,d", maxOf(0, n))

fun formatCountdown(ms: Long): String {
    val total = maxOf(0, floor(ms / 1000.0).toLong())
    val h = total / 3600
    val m = (total % 3600) / 60
    if (h > 0) return "${h}h ${m}m"
    val s = total % 60
    if (m > 0) return "${m}m ${s}s"
    return "${s}s"
}
